randomgen/src/distributions/distributions.c

```diff
@@ -1220,42 +1222,86 @@ static NPY_INLINE uint64_t bounded_uint64(brng_t *brng_state, uint64_t off,
   return off + val;
 }
 
-uint64_t random_bounded_uint64(brng_t *brng_state, uint64_t off, uint64_t rng,
+uint64_t random_bounded_uint64(brng_t *brng_state,
+                               uint64_t off, uint64_t rng,
                                uint64_t mask) {
   return bounded_uint64(brng_state, off, rng, mask);
 }
 
-static NPY_INLINE uint32_t bounded_uint32(brng_t *brng_state, uint32_t off,
-                                          uint32_t rng, uint32_t mask) {
+static NPY_INLINE uint32_t bounded_masked_uint32(brng_t *brng_state,
+                                                 uint32_t off, uint32_t rng,
+                                                 uint32_t mask) {
   /*
    * The buffer and buffer count are not used here but are included to allow
    * this function to be templated with the similar uint8 and uint16
    * functions
    */
-
   uint32_t val;
-  if (rng == 0)
-    return off;
 
   while ((val = (next_uint32(brng_state) & mask)) > rng)
     ;
   return off + val;
 }
 
-uint32_t random_buffered_bounded_uint32(brng_t *brng_state, uint32_t off,
-                                        uint32_t rng, uint32_t mask, int *bcnt,
-                                        uint32_t *buf) {
+static NPY_INLINE uint32_t bounded_lemire_uint32(brng_t *brng_state,
+                                                 uint32_t off, uint32_t rng) {
+  /*
+   * Uses Lemire's algorithm - https://arxiv.org/abs/1805.10941
+   *
+   * The buffer and buffer count are not used here but are included to allow
+   * this function to be templated with the similar uint8 and uint16
+   * functions
+   */
+  const uint32_t rng_excl = rng + 1;
+  uint64_t m;
+  uint32_t leftover;
+
+  if (rng == 0xFFFFFFFF) {
+    return next_uint32(brng_state);
+    /* ToDo: Move this code to caller to prevent this check on each call when generating arrays of numbers. */
+  }
+
+  /* Generate a scaled random number. */
+  m = ((uint64_t)next_uint32(brng_state)) * rng_excl;
```

Comment:
It looks pretty easy to implement a version for 64/128 using manual carry.  Will need #def paths for when __uint128 is available and when it isn't (and then use manual carry).  I think what is needed is virtually identical to what is here:

https://github.com/bashtage/randomgen/blob/master/randomgen/src/pcg64/pcg64.h#L95

Reply:
I agree. The code you show only uses 128bit on Windows though. One will have to add the #def paths for when __uint128 is available as proposed above under Linux.

Reply:
The __uint128 step is already in pcg64.  That part is easy since it is identical to what you have here only using the correct types.

Reply:
ok, I found the `#ifdef PCG_EMULATED_128BIT_MATH`. I'll do it similar to that. Doing what you propose in https://github.com/bashtage/randomgen/pull/29#discussion_r223180392 will be more invasive, but probably well worth it if the platform doesn't support the uint128.